kworkflow/kworkflow
# Integrate `kw_bd` to the project and add migration script

This PR is a continuation of the work made by @kwy95 and @magalilemes in [**src: db: introduce db model and migration script**](https://github.com/kworkflow/kworkflow/pull/531) (thanks for the excellent work!). Its purpose is to integrate `kw_bd`, the implementation of a real database management system that uses SQLite underneath, into the project. In other words, wherever metadata files are being used to simulate a `DB` or wherever `DB`'s functionalities are advantageous, this integration should happen. Also, this PR adds a script to migrate old metadata files into the `DB`.

**[TODOs]** At the moment this PR is just a draft as there are several missing parts of it:
- ~~Update `kernel-config-manager` test suite (integration is done)~~
- ~~FInish integration with `build` and `deploy` (statistics)~~
- ~~Integrate with `pomodoro`~~
- ~~Integrate with `report`~~
- ~~Fix migration script (migrate the same legacy user data only once)~~

**[TESTING THIS PR]**
If you wish to test this PR (thanks a lot!), below are the steps and the assertions you should make. The TLDR is that just by pulling this PR, installing the PR version of kw and asserting the migration is consistent using `kw report` and `kw kernel-config-manager` is of great help! The steps below are a detailed script to fully test the PR (besides automated tests).

1. It is **REALLY IMPORTANT** to make the backup of your kw personal data. Either use `kw backup <path_to_backup>` or `cp -r <kw_data_dir_path> <path_to_backup>`
2. Have the `unstable` branch updated with `git pull origin unstable`
3. Checkout to the `unstable` branch with `git checkout unstable`
4. Pull this PR locally with `git pull origin pull/836/head:<name_of_branch>`
5. Checkout to the local PR branch with `git switch <name_of_branch>`
6. Install this version of kw with `./setup.sh --install`. You should see a warning like `Creating database: <kw_data_dir_path>/kw.db` and no errors
7. Assert that the file `<kw_data_dir_path>/kw.db` has been created
8. Assert that in your `<kw_data_dir_path>`, the directories `pomodoro` and `statistics` are renamed to `legacy_pomodoro` and `legacy_statistics`, respectively. For the kernel configs, assert that in your `<kw_data_dir_path>/configs`, the directories `configs` and `metadata` are renamed to `legacy_configs` and `legacy_metadata` and that the actual kernel config files (the kernel `.config`s) are in `<kw_data_dir_path>/configs`
9. Run `kw report <period of time>` of a period of time that you can assert the output
10. Experiment with `kw build`, `kw deploy`, `kw deploy --list`, `kw deploy --uninstall` or `kw deploy --modules-deploy` and assert through `kw report --statistics` their addition
11. Experiment with `kw pomodoro --set-timer <timebox>` and assert through `kw report --pomodoro` their addition. Also use `kw pomodoro --check-timer` to see if the active timeboxes are displayed correctly
12. Experiment with `kw kernel-config-manager` and assert that the feature works as intended
13. Reinstall this version of kw with `./setup.sh --install`. You should **NOT** see a warning like `Creating database: <kw_data_dir_path>/kw.db` (also, no errors should happen)
14. Assert that `<kw_data_dir_path>` hasn't changed and that there is no duplicated data. You could use `sqlite3 <kw_data_dir_path>/kw.db` and check that the tables are consistent (e.g. `SELECT * FROM statistics_report;`) or juts use `kw report`
15. If any corruption of your data has happened, restore your backup (recommend deleting the contents of `<kw_data_dir_path>` and running `kw backup --restore <path_to_backup> --force`)
16. THANK YOU SO MUCH FOR TESTING THIS PR!
